File: leetcode/linked_list/24.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        // // Method 1: single loop
        // if ( !(head) || !(head->next) ) {
        //     return head;
        // }
        // ListNode *dummy_head = new ListNode(0, head);
        // ListNode *i = dummy_head, *j = head, *k = head->next;
        // while ( 1 ) {
        //     j->next = k->next;
        //     k->next = j;
        //     i->next = k;
        //     if ( j->next && j->next->next ) {
        //         i = i->next->next;
        //         j = j->next;
        //         k = j->next;
        //     } else {
        //         break;
        //     }
        // }
        // return dummy_head->next;

        // Method 2: recursion
        if ( !head || !(head->next) ) {
            return head;
        }
        ListNode *i = head;
        ListNode *j = head->next;
        ListNode *k = head->next->next;
        j->next = i;
        i->next = swapPairs(k);
        return j;
    }
};

File: leetcode/linked_list/24.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var swapPairs = function(head) {
  // // Method 1: recursion
  // if ( head === null ) {
  //     return null;
  // }
  // function recurSwap(a, b) {
  //     if ( a === null ) {
  //         return null;
  //     }
  //     if ( b === null ) {
  //         return a;
  //     }
  //     a.next = recurSwap(b.next, b.next === null ? null : b.next.next);
  //     b.next = a;
  //     return b;
  // }
  // return recurSwap(head, head.next);

  // Method 2: iteration
  if ( head === null || head.next === null ) {
      return head;
  }
  let a = head;
  let b = head.next;
  let c = new ListNode(0, head);
  let dummyHead = new ListNode(0, c);

  while ( 1 ) {
      // console.log(a.val, b.val);
      a.next = b.next;
      b.next = a;
      c.next = b;

      if ( a.next === null || a.next.next === null ) {
          break;
      }
      c = a;
      a = a.next;
      b = a.next;
  }
  
  return dummyHead.next.next;
};